src/utils/middleware.ts

```diff
@@ -0,0 +1,205 @@
+import type {
+  JSONRPCMessage,
+  JSONRPCRequest,
+  JSONRPCResponse,
+  JSONRPCResponseResult,
+  MiddlewareFactory,
+} from '../types';
+import { TransformStream } from 'stream/web';
+import { JSONParser } from '@streamparser/json';
+import * as rpcUtils from './utils';
+import { promise } from './utils';
+import * as rpcErrors from '../errors';
+
+/**
+ * This function is a factory to create a TransformStream that will
+ * transform a `Uint8Array` stream to a JSONRPC message stream.
+ * The parsed messages will be validated with the provided messageParser, this
+ * also infers the type of the stream output.
+ * @param messageParser - Validates the JSONRPC messages, so you can select for a
+ *  specific type of message
+ * @param bufferByteLimit - sets the number of bytes buffered before throwing an
+ *  error. This is used to avoid infinitely buffering the input.
+ */
+function binaryToJsonMessageStream<T extends JSONRPCMessage>(
+  messageParser: (message: unknown) => T,
+  bufferByteLimit: number = 1024 * 1024,
+): TransformStream<Uint8Array, T> {
+  const parser = new JSONParser({
+    separator: '',
+    paths: ['$'],
+  });
+  let bytesWritten: number = 0;
+
+  return new TransformStream<Uint8Array, T>({
+    flush: async (controller) => {
+      const waitP = promise();
+      parser.onEnd = () => waitP.resolveP();
+      parser.end();
+      await waitP.p;
+    },
+    start: (controller) => {
+      parser.onValue = (value) => {
+        const jsonMessage = messageParser(value.value);
+        const jsonMessageString = JSON.stringify(jsonMessage);
+        const jsonMessageWithNewline = jsonMessageString + '\n';
+        const jsonMessageObject: T = JSON.parse(jsonMessageWithNewline) as T;
```

Comment:
i don't think this is a good idea, JSON.parse(JSON.stringify(...)) is slow

Reply:
Already had a conversation with Aditya, this solution is not correct. It should be on the parser side.

Reply:
WIP